vmm/src/device_manager.rs

```diff
@@ -723,62 +834,111 @@ impl DeviceManager {
+    }
 
-        if cfg!(feature = "pci_support") {
-            device_manager.add_pci_devices(virtio_devices.clone())?;
-        } else if cfg!(feature = "mmio_support") {
-            device_manager.add_mmio_devices(virtio_devices.clone(), &legacy_interrupt_manager)?;
-        }
+    pub fn create_devices(&mut self) -> DeviceManagerResult<()> {
+        let mut virtio_devices: Vec<(VirtioDeviceArc, bool, String)> = Vec::new();
 
-        device_manager.virtio_devices = virtio_devices;
+        let ioapic = self.add_ioapic()?;
 
-        let device_manager = Arc::new(Mutex::new(device_manager));
+        // Now we can create the legacy interrupt manager, which needs the freshly
+        // formed IOAPIC device.
+        let legacy_interrupt_manager: Arc<
+            dyn InterruptManager<GroupConfig = LegacyIrqGroupConfig>,
+        > = Arc::new(KvmLegacyUserspaceInterruptManager::new(ioapic));
 
         #[cfg(feature = "acpi")]
-        address_manager
+        self.address_manager
             .allocator
             .lock()
             .unwrap()
-            .allocate_io_addresses(Some(GuestAddress(0xae00)), 0x10, None)
+            .allocate_io_addresses(Some(GuestAddress(0x0a00)), 0x18, None)
             .ok_or(DeviceManagerError::AllocateIOPort)?;
 
         #[cfg(feature = "acpi")]
-        address_manager
+        self.address_manager
             .io_bus
             .insert(
-                Arc::clone(&device_manager) as Arc<Mutex<dyn BusDevice>>,
-                0xae00,
-                0x10,
+                Arc::clone(&self.memory_manager) as Arc<Mutex<dyn BusDevice>>,
+                0xa00,
+                0x18,
             )
             .map_err(DeviceManagerError::BusError)?;
 
-        Ok(device_manager)
+        self.add_legacy_devices(
+            self.reset_evt
+                .try_clone()
+                .map_err(DeviceManagerError::EventFd)?,
+        )?;
+
+        #[cfg(feature = "acpi")]
+        {
+            self.ged_notification_device = self.add_acpi_devices(
+                &legacy_interrupt_manager,
+                self.reset_evt
+                    .try_clone()
+                    .map_err(DeviceManagerError::EventFd)?,
+                self.exit_evt
+                    .try_clone()
+                    .map_err(DeviceManagerError::EventFd)?,
+            )?;
+        }
+
+        self.console = self.add_console_device(&legacy_interrupt_manager, &mut virtio_devices)?;
+
+        #[cfg(any(feature = "pci_support", feature = "mmio_support"))]
+        virtio_devices.append(&mut self.make_virtio_devices()?);
+
+        if cfg!(feature = "pci_support") {
+            self.add_pci_devices(virtio_devices.clone())?;
+        } else if cfg!(feature = "mmio_support") {
+            self.add_mmio_devices(virtio_devices.clone(), &legacy_interrupt_manager)?;
+        }
+
+        self.virtio_devices = virtio_devices;
+
+        Ok(())
+    }
+
+    fn state(&self) -> DeviceManagerState {
```

Comment:
Could we embed `DeviceManagerState` into `DeviceManager` rather than the individual values?

Reply:
This is gonna create a new level of indirection inside the device manager to access the device tree and the device id counter.
This `DeviceManagerState` structure is not really meant to be maintained at all time, the point is just to create it when we need to save the current state.